js/ui/diagram/ui.diagram.js

```diff
@@ -89,26 +94,118 @@ class Diagram extends Widget {
         this._diagramInstance = new DiagramControl();
         this._diagramInstance.onChanged = this._raiseDataChangeAction.bind(this);
     }
+    _refreshDataSources() {
+        this._beginUpdateDiagram();
+        this._refreshNodesDataSource();
+        this._refreshEdgesDataSource();
+        this._endUpdateDiagram();
+    }
+    _refreshNodesDataSource() {
+        if(this._nodesOption) {
+            this._nodesOption._disposeDataSource();
+            delete this._nodesOption;
+            delete this._nodes;
+        }
+        if(this.option("nodes.dataSource")) {
+            this._nodesOption = new NodesOption(this);
+            this._nodesOption.option("dataSource", this.option("nodes.dataSource"));
+            this._nodesOption._refreshDataSource();
+        }
+    }
+    _refreshEdgesDataSource() {
+        if(this._edgesOption) {
+            this._edgesOption._disposeDataSource();
+            delete this._edgesOption;
+            delete this._edges;
+        }
+        if(this.option("edges.dataSource")) {
+            this._edgesOption = new EdgesOptions(this);
+            this._edgesOption.option("dataSource", this.option("edges.dataSource"));
+            this._edgesOption._refreshDataSource();
+        }
+    }
+    _getDiagramData() {
+        let value;
+        const { DiagramCommand } = getDiagram();
+        this._diagramInstance.commandManager.getCommand(DiagramCommand.Export).execute(function(data) { value = data; });
+        return value;
+    }
+    _setDiagramData(data, keepExistingItems) {
+        const { DiagramCommand } = getDiagram();
+        this._diagramInstance.commandManager.getCommand(DiagramCommand.Import).execute({
+            data,
+            keepExistingItems
+        });
+    }
+    _nodesDataSourceChanged(nodes) {
+        this._nodes = nodes;
+        this._bindDiagramData();
+    }
+    _edgesDataSourceChanged(edges) {
+        this._edges = edges;
+        this._bindDiagramData();
+    }
+    _bindDiagramData() {
+        if(this._updateDiagramLockCount || !this._isBindingMode()) return;
+
+        const { DiagramCommand } = getDiagram();
+        const data = {
+            nodeDataSource: this._nodes,
+            edgeDataSource: this._edges,
+            nodeMapping: {
+                idField: this.option("nodes.idExpr"),
+                textField: this.option("nodes.textExpr"),
+                typeField: this.option("nodes.typeExpr")
+            },
+            edgeMapping: {
+                idField: this.option("edges.idExpr"),
+                fromField: this.option("edges.fromExpr"),
+                toField: this.option("edges.toExpr")
+            },
+            layoutType: this._getDataLayoutType()
+        };
+        this._diagramInstance.commandManager.getCommand(DiagramCommand.BindDocument).execute(data);
+    }
+    _getDataLayoutType() {
+        const { DataLayoutType } = getDiagram();
+        switch(this.option("layout")) {
+            case "tree":
```

Comment:
You may remove the `case "tree":` code. Default case will handle this situation too.